app/controllers/api/alert_definitions_controller.rb

```diff
@@ -1,25 +1,39 @@
 module Api
   class AlertDefinitionsController < BaseController
-    REQUIRED_FIELDS = %w(description db expression options).freeze
+    REQUIRED_FIELDS = %w(description db options).freeze
 
     before_action :set_additional_attributes
 
     def create_resource(type, id, data = {})
       assert_id_not_specified(data, type)
       assert_all_required_fields_exists(data, type, REQUIRED_FIELDS)
+      if expression_values(data).size != 1
+        raise BadRequestError, "Must specify exactly one expression for creating a new #{type} resource"
+      end
       begin
-        data["expression"] = MiqExpression.new(data["expression"])
+        create_miq_expression(data)
+        data["options"].deep_symbolize_keys!
         data["enabled"] = true if data["enabled"].nil?
-        super(type, id, data).serializable_hash.merge("expression" => data["expression"])
+        super(type, id, data).serializable_hash.merge("expression" => expression_values(data).first)
       rescue => err
         raise BadRequestError, "Failed to create a new alert definition - #{err}"
       end
     end
 
     def edit_resource(type, id = nil, data = {})
       raise BadRequestError, "Must specify an id for editing a #{type} resource" unless id
+      exp_values = expression_values(data)
+      if exp_values.size > 1
+        raise BadRequestError, "Must not specify more than one expression for editing a #{type} resource"
+      elsif exp_values.size == 1
+        if data["hash_expression"]
+          data["miq_expression"] = nil
```

Comment:
So as we discussed with @zakiva this handles an edge case where a user wants to replace a hash expression with a miq expression or vice versa. Honestly I don't know if we should support that